### PR TITLE
water mist

### PR DESCRIPTION
**Pull request by pinkysnowman**
Added water mist particle effects to enhance waterfalls and other water features.
screenshot_20150911_192515
screenshot_20150911_192504

### CODE REVIEW

**Comment by VanessaE** [comment]
:+1:

**Comment by VanessaE** [comment]
Some tweaks to make it even better:  http://pastebin.com/t4y28SDQ

adjusted the expiry times on the two spawner calls, and increased the particle count on the second (just flowing down a hill) one.

**Comment by kilbith** [comment]
Nice, but not until we have the Irrlicht particles for obvious performance strangling.

File: mods/default/functions.lua [hunk 1]
```diff
@@ -346,3 +346,59 @@ minetest.register_abm({
...
+--
+-- Water mist 
+--
+
+minetest.register_abm({
+	nodenames = {"default:water_flowing"},
+	interval = 5,
+	chance = 1,
+	action = function(pos, node, active_object_count, active_object_count_wider)
+		if pos.y < 0 then return end
```
Comment (PilzAdam) [inline mods/default/functions.lua]: Why disable it underground?

**Comment by pinkysnowman** [comment]
it is disabled as i have found some lakes to have only one layer of water_source on older builds and also caves under bodies of water can cause massive amounts of lag from the sheer amount of flowing water

**Comment by PilzAdam** [comment]
Adding water_source to neighbors may help with performance.

**Comment by pinkysnowman** [comment]
unfortunately i t cant be added, as it has some particles on flowing water that's just flowing over land.

**Comment by VanessaE** [comment]
Yep, I can confirm the performance issues.   :frown:  

Too bad, too, as this PR adds a very nice effect.

**Comment by pinkysnowman** [comment]
i have removed effects from flowing water and left the waterfall effects. I am still testing code for performance and will update asap!

**Comment by paramat** [comment]
I feel this is best as a mod since it's a special visual feature that's triggered in rare circumstances.
An ABM on flowing water creates a processing load that in most worlds, most of the time, is doing nothing, i don't think this should be forced on everyone by default, for example i would rather not use it to improve performance.
Particles in Minetest are currently heavy, this feature creates 33 per node with collision detection.
This looks good under a waterfall, but in many situations, like a small stream of water running into a pool, spray like this would be unsuitable.

**Comment by HybridDog** [comment]
you could make semi transparent particles

**Comment by paramat** [comment]
Translucent textures are very buggy in Minetest due to no z-sorting, not recommended.

**Comment by C1ffisme** [comment]
This looks pretty awesome, but as said before it can be laggy.

The particles could be a bit less bright, I think.

And a few of the particles should look a bit more like drips of water... That way it looks like water is splashing everywhere.

**Comment by PilzAdam** [comment]
:-1: from me; I agree with @paramat.

File: mods/default/functions.lua [hunk 2]
```diff
@@ -346,3 +346,40 @@ minetest.register_abm({
...
+	neighbors = {"default:water_source"},
+	interval = 5,
+	chance = 1,
+	action = function(pos, node, active_object_count, active_object_count_wider)
+		local posa = {x=pos.x, y=pos.y+1, z=pos.z}
+		local posu = {x=pos.x, y=pos.y-1, z=pos.z}
+		local na =  minetest.get_node(posa).name
+		local nu =  minetest.get_node(posu).name
+		if na == "default:water_flowing" then
+			if nu == "default:water_source" then
```
Comment (ShadowNinja) [inline mods/default/functions.lua]: This indentation gets very deep.  You can reduce it by two levels by combining and inverting the conditionals like so:
```Lua
if na != "default:water_flowing" or nu != "default:water_source" then
        return
end
minetest.add_particlespawner...
```

**Comment by ShadowNinja** [comment]
@paramat, @PilzAdam: The texture appears to be opaque.

File: mods/default/functions.lua (hunk 2 above)
Comment (HybridDog) [inline mods/default/functions.lua]: What does ```!=``` do?

Comment (pinkysnowman) [inline mods/default/functions.lua]: != is doesn't equal in c/c++ i think SN  meant ~= but same thing just
different lang

Comment (ShadowNinja) [inline mods/default/functions.lua]: You could do that too and increase the performance by a tad.

**Comment by paramat** [comment]
Sure, i was responding to HybridDog's comment.

**Comment by paramat** [comment]
:-1:

**Comment by sfan5** [comment]
Very nice effect but it's too laggy/buggy in it's current state :-1:

**Comment by paramat** [comment]
Closing due to 3 :-1:

**Pull request closed without merging**